### DISCUSSION TITLE
TRPC client is null when using monorepo

### DISCUSSION THREAD

**Discussion by Pythonista7** [trpc/trpc#2410] [Q&A] [answered]
I am trying to use trpc in a turbo repo structured like:
```
/
|
|-- apps/
|       |-- webfront/<nextjs project>
|
|-- packages/server/<trpc code>
``` 

Inside `<nextjs project's api folder>` i have `[trpc].ts` as :
```js
// /home/ash/apps/webfront/src/pages/api/trpc/[trpc].ts
import * as trpcNext from '@trpc/server/adapters/next'
import { createContext } from 'server/createContext'
import appRouter from 'server/routes/route'

export default trpcNext.createNextApiHandler({
	router: appRouter,
	createContext,
	onError({ error }) {
		console.error(error)
	},
})
``` 
and the file that exports trpc types from the server is :
```js
// /home/ash/packages/server/trpc.ts
/* eslint-disable @typescript-eslint/naming-convention */
import { createReactQueryHooks } from '@trpc/react'
import type { inferProcedureInput, inferProcedureOutput } from '@trpc/server'
import { AppRouter } from './routes/route'

const trpc = createReactQueryHooks<AppRouter>()

/**
 * This is a helper method to infer the output of a query resolver
 * @example type HelloOutput = inferQueryOutput<'hello'>
 */
export type inferQueryOutput<
	TRouteKey extends keyof AppRouter['_def']['queries']
> = inferProcedureOutput<AppRouter['_def']['queries'][TRouteKey]>

export type inferQueryInput<
	TRouteKey extends keyof AppRouter['_def']['queries']
> = inferProcedureInput<AppRouter['_def']['queries'][TRouteKey]>

export type inferMutationOutput<
	TRouteKey extends keyof AppRouter['_def']['mutations']
> = inferProcedureOutput<AppRouter['_def']['mutations'][TRouteKey]>

export type inferMutationInput<
	TRouteKey extends keyof AppRouter['_def']['mutations']
> = inferProcedureInput<AppRouter['_def']['mutations'][TRouteKey]>

export default trpc

```

But when i try using the trpc query in a nextjs page (not-ssr'd) , I run into the error:

```
TypeError: Cannot read property 'client' of null
```
Error stack:
![image](https://user-images.githubusercontent.com/36104244/183663863-15734ef7-742f-4958-aceb-e12d7cb46d51.png)

```
processChild
file:///home/ash/node_modules/react-dom/cjs/react-dom-server.node.development.js (3353:14)
resolve
file:///home/ash/node_modules/react-dom/cjs/react-dom-server.node.development.js (3270:5)
ReactDOMServerRenderer.render
file:///home/ash/node_modules/react-dom/cjs/react-dom-server.node.development.js (3753:22)
ReactDOMServerRenderer.read
file:///home/ash/node_modules/react-dom/cjs/react-dom-server.node.development.js (3690:29)
Object.renderToString
file:///home/ash/node_modules/react-dom/cjs/react-dom-server.node.development.js (4298:27)
renderDocument
file:///home/ash/node_modules/next/dist/server/render.js (812:47)
Object.renderToHTML
file:///home/ash/node_modules/next/dist/server/render.js (893:34)
runMicrotasks
<anonymous>
```

Terminal Dump:
```
error - ../../packages/server/node_modules/@trpc/react/dist/trpc-react.cjs.dev.js (160:0) @ Object.useQuery
webfront:dev: error - TypeError: Cannot read property 'client' of null
webfront:dev:     at Object.useQuery (webpack-internal:///../../packages/server/node_modules/@trpc/react/dist/trpc-react.cjs.dev.js:160:31)
webfront:dev:     at FlickTab (webpack-internal:///./src/components/dashboard/FlickTab.tsx:54:132)
```
TRPC Version:
```
"@trpc/react": "^9.26.2",
"@trpc/server": "^9.25.3"
```
I wanted to have the trpc in a seperate folder in `packages` because there could be a need to use the same queries in different apps (i.e: maybe a new nextjs app in the monorepo) .

**Comment by juliusmarminge** [comment, marked as answer]
Not at my computer so can't check the exact code but take a look at this for reference, it has similar setup as you are describing: https://github.com/juliusmarminge/turbo-expo-next-starter

**Reply by Pythonista7** [reply]
Thanks so much for this @juliusmarminge  ! I got it working :)

I had been exporting this:
```js
export const trpc = createReactQueryHooks<AppRouter>();
```
from `/packages/server/trpc.ts` and using it in the nextjs project. This seems to have been the issue, once i moved the file from `/packages/server/trpc.ts` to `apps/webfront/server/trpc.ts` (i.e: inside the nextjs folder) everything seems to be fine. But i am curious as to why this is the case...

**Reply by juliusmarminge** [reply]
If I'm not mistaken, you should be able to create like a barrelfile in the next repo if you rather want that, zart has that:

https://github.com/trpc/zart/blob/0daf1dbc855825a950523e7b50abe6cd619aa2a7/apps/nextjs/src/utils/trpc.ts#L1

which is just importing from a shared react package: 

https://github.com/trpc/zart/blob/main/packages/react/trpc.tsx

**Comment by sachinraja** [comment]
Can we see your `_app.ts` file? My guess is you have two different versions of `@trpc/react` installed so the `useContext` for the one in the nextjs package is just returning null since `withTRPC` uses a different instance of `@trpc/react`.
